### DISCUSSION TITLE
EMS-ESP 3.6 on Lolin c3 Mini

### DISCUSSION THREAD

**Discussion by martinSezman** [emsesp/EMS-ESP32#1241] [General]
Hi,
I don't know what I doing wrong, but the ESMESP v3.6 don't boot on esp32 lolin c3 mini. Board is correct - verified by install tasmota and esphome.

EMS-ESP-3_6_0-dev_12-LOLIN_C3_MINI.bin firmware compiled and builded without issues myself by command "pio run -e lolin_c3_mini" on Debian 12.

Upload command on macOS:
esptool.py --chip esp32c3 --port /dev/cu.usbmodem14601 --baud 921600 --before=default_reset --after=hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size 4MB 0x8000 partitions.bin 0x0 bootloader.bin 0xe000 boot_app0.bin 0x10000 EMS-ESP-3_6_0-dev_12-LOLIN_C3_MINI.bin

on serial are following messages in the loop:
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0xd (SPI_FAST_FLASH_BOOT)
Saved PC:0x403d10fc
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd6100,len:0x38c
load:0x403ce000,len:0x6ac
load:0x403d0000,len:0x24e4
entry 0x403ce000

Any ideas, suggestions?

Best regards.

**Comment by proddy** [comment]
maybe try the latest code from dev and rebuild. I'm not at home this week so can't test unfortunately.

**Reply by martinSezman** [reply]
today builded from latest dev EMS-ESP-3_6_0-dev_14-LOLIN_C3_MINI.bin
and result is exacly the same.

**Reply by MichaelDvP** [reply]
This is on all chips if you have updated to the latest framework-arduinoespressif32.
Use `platform = espressif32@6.3.1` in your `pio_local.ini`, see #1242

**Reply by martinSezman** [reply]
Thanks, it works !

**Reply by MichaelDvP** [reply]
If you have any idea what causes the bootloop in platform 6.3.2, please share in #1242. 
It's the same RTC watchdog reset for esp32, esp32c3 and esp32s3. Uploading to the second partition do not cause bootloop, only boot to the old partition. Serial log does not report any error. Seems the bootloader does not start the 6.3.2 compiled code.

**Reply by proddy** [reply]
I have some time tomorrow so will investigate

**Reply by proddy** [reply]
added comments to https://github.com/emsesp/EMS-ESP32/issues/1242

**Comment by peter--s** [comment]
Hi!
@MichaelDvP is this still relevant today? With https://github.com/emsesp/EMS-ESP32/issues/1242 having been closed, my guess was that (both) using the latest framework (and flashing released binaries) should work without issues.

Background:
I am willing to move from my old ESP32 to a much smaller ESP32-C3 Super Mini.
1. After flashing the prebuilt EMS-ESP-3_6_5-ESP32.bin using ESP-Flasher v1.0.3 to the ESP32C3 SuperMini resulted in a boot loop, I ended up here.
2. I then tried flashing the above binary using esptool.py, which indicated, that partitions.bin, which I downloaded from https://github.com/emsesp/EMS-ESP32/EMS-ESP32/scripts as suggested in https://github.com/emsesp/EMS-ESP32/issues/14, was probably not compiled for ESP32C3.
3. I then downloaded partitions_esp32c3_4MB.bin from https://github.com/emsesp/EMS-ESP-Flasher/tree/main/partitions and bootloader_esp32c3.bin and boot_app0.bin from https://github.com/emsesp/EMS-ESP-Loader/tree/main to solve that issue.
4. After that, esptool only complained that EMS-ESP-3_6_5-ESP32.bin was not compiled for ESP32C3:
```
./Library/Python/3.11/bin/esptool.py --chip auto --port /dev/cu.usbmodem14201 --baud 921600 --before=default_reset --after=hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size 4MB 0x8000 Downloads/partitions_esp32c3_4MB.bin 0x0 Downloads/bootloader_esp32c3.bin 0xe000 Downloads/boot_app0.bin 0x10000 Downloads/EMS-ESP-3_6_5-ESP32.bin 
esptool.py v4.7.0
Serial port /dev/cu.usbmodem14201
Connecting....
Detecting chip type... ESP32-C3
Chip is ESP32-C3 (QFN32) (revision v0.4)
Features: WiFi, BLE, Embedded Flash 4MB (XMC)
Crystal is 40MHz
MAC: xx:xx:xx:xx:xx:xx
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Unexpected chip id in image. Expected 5 but value was 0. Is this image for a different chip model?

A fatal error occurred: Downloads/EMS-ESP-3_6_5-ESP32.bin is not an ESP32-C3 image. Use --force to flash anyway.
```
The offered option (to ignore that and flash the binary) resulted in a boot loop again:
```
./Library/Python/3.11/bin/esptool.py --chip auto --port /dev/cu.usbmodem14201 --baud 921600 --before=default_reset --after=hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size 4MB 0x8000 Downloads/partitions_esp32c3_4MB.bin 0x0 Downloads/bootloader_esp32c3.bin 0xe000 Downloads/boot_app0.bin 0x10000 Downloads/EMS-ESP-3_6_5-ESP32.bin --force
esptool.py v4.7.0
Serial port /dev/cu.usbmodem14201
Connecting....
Detecting chip type... ESP32-C3
Chip is ESP32-C3 (QFN32) (revision v0.4)
Features: WiFi, BLE, Embedded Flash 4MB (XMC)
Crystal is 40MHz
MAC: xx:xx:xx:xx:xx:xx
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x00000000 to 0x00003fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x0019dfff...
Compressed 3072 bytes to 127...
Wrote 3072 bytes (127 compressed) at 0x00008000 in 0.0 seconds (effective 498.6 kbit/s)...
Hash of data verified.
Warning: Image file at 0x0 is protected with a hash checksum, so not changing the flash frequency setting. Use the --flash_frequency=keep option instead of --flash_frequency=40m in order to remove this warning, or use the --dont-append-digest option for the elf2image command in order to generate an image file without a hash checksum
Compressed 12944 bytes to 9297...
Wrote 12944 bytes (9297 compressed) at 0x00000000 in 0.2 seconds (effective 428.4 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 679.4 kbit/s)...
Hash of data verified.
Compressed 1629376 bytes to 1136450...
Wrote 1629376 bytes (1136450 compressed) at 0x00010000 in 13.1 seconds (effective 995.3 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0xd (SPI_FAST_FLASH_BOOT)
Saved PC:0x403d10fc
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd6100,len:0x438
load:0x403ce000,len:0x918
load:0x403d0000,len:0x24e4
entry 0x403ce000
```
5.  Since I am neither confident that **either of the released binaries are expected to function on ESP32C3**, nor that **3.6.5 (dating from March 2023(?) and yet reporting Tasmota Arduino v2.0.14 / ESP-IDF 4.4.6.240105 on my old ESP32) is not affected by [#1242](https://github.com/emsesp/EMS-ESP32/issues/1242)**, I would like to compile ems-esp myself. 
So I am wondering what I need to do with pio_local.ini for ESP32C3 to create a compatible binary and avoid possible toolchain issues.

Thanks&Regards
Peter

**Reply by MichaelDvP** [reply]
We don't have precompiled binaries for esp32c3. You need to compile yourself. You can use the c3mini env, this also should work on the supermini, but you'll need customboard profile. 
Actually the 3.7.0-dev cannot be compiled, because there is no modules.lib for the c3. But you can compile the 3.6.5 release. But i think this will be solved soon.
But i think moving to a chip with dual core, more flash and ram is much better, the c3 is not a god solution.

**Reply by peter--s** [reply]
OK, here is what I did to get a working binary for ESP32C3:

1. Fire up Gitpod: https://gitpod.io/#https://github.com/emsesp/EMS-ESP32
2. Get PlatformIO using the terminal:
```
curl -fsSL https://raw.githubusercontent.com/platformio/platformio-core-installer/develop/get-platformio.py>get-platformio.py
python3 get-platformio.py
mkdir -p /usr/local/bin
sudo ln -s ~/.platformio/penv/bin/piodebuggdb /usr/local/bin/piodebuggdb
sudo ln -s ~/.platformio/penv/bin/pio /usr/local/bin/pio
sudo ln -s ~/.platformio/penv/bin/platformio /usr/local/bin/platformio
```
3. Build EMS-ESP using martinSezman's approach instead of following https://emsesp.github.io/docs/Building/:
```
pio run -e lolin_c3_mini
```
4. Download the binary `build/firmware/EMS-ESP-3_6_5-LOLIN_C3_MINI.bin`
5. Flash using the command line and the other C3 .bin files from above.
6. Enjoy EMS-ESP on my tiny ESP32-C3. :-)

The GUI reports the same SDK as my old ESP32 flashed with the binary from the 3.6.5 release quoted above.
My conclusion is that my boot loop did not have much to do with the toolchain issue mentioned here.
```
show system
System:
 Version: 3.6.5
 Language: en
 Board profile: C3MINI
 Uptime: 000+00:20:17.804
 SDK version: 4.4.6.240105
 CPU frequency: 160 MHz
 Free heap/Max alloc: 181 KB / 155 KB
 App used/free: 1574 KB / 410 KB
 FS used/free: 24 KB / 40 KB
```
I hope somebody will find this useful, though. I shall head over to the physical modifications of the gateway pcb.

**Reply by peter--s** [reply]
Thanks, @MichaelDvP!
You must have answered while I was editing my reply. Well, I will attempt to cram the interface behind my RC30, so size matters. I will see how the C3 performs. I expect little need for updates, since my heating is old and I have been happy with EMS-ESP stability for the last years. So if 3.6.5 works reasonably well, it will stay.

**Reply by entwickler32** [reply]
HI @peter--s 
when trying the mentioned build steps, the build process is running through successfully.
But when i try to flash the firmware it runs in a bootloop:entry 0x403ce000.
The device is flashed with partitions_esp32c3_4MB.bin from https://github.com/emsesp/EMS-ESP-Flasher/tree/main/partitions and bootloader_esp32c3.bin and boot_app0.bin from https://github.com/emsesp/EMS-ESP-Loader/tree/main and finally my compiled firmware.
When flashed wiht the EMS-ESP-Loader.bin the System starts up and is working. But its only the loader. Upgrading the firmware with the esp hosted web-updater is not working, too.
When using the official flashing tool (and not esptool) it complains about a wrong magic byte (magic byte=EF, should be E9)#.
Is there any hint to get the actual firmware working on the esp32c3mini?

**Reply by bbqkees** [reply]
AFAIK the C3 Mini is pin-compatible with the S3 Mini and others in this form factor.
Almost everyone is using either an ESP32 or ESP32-S3 module for EMS-ESP, do not really see anyone using the C3.
So in the end it is probably easier to swap that C3 for an S3 board and get the support and reliability with the S3 platform and pre-compiled binaries.

**Reply by bbqkees** [reply]
And to add to it, the C3 only has 400KB of SRAM, which is even less than the already limiting 520KB of the ESP32.
Just get a board with a S3 and some PSRAM on it, flash the bin file from this repository and you can run it for years to come.

**Reply by entwickler32** [reply]
Thank you, i already replaced the c3 with an esp32. The board i used is much bigger than the c3mini, but the whole thing is mounted on the boiler, so it does not matter.
Is there any roadmap which points that the normal esp32wroom would not be enogh in the future?
